libtiledbsoma/src/soma/soma_array.cc

```diff
@@ -966,6 +966,235 @@ bool SOMAArray::_cast_column_aux<bool>(
+    // dictionary's indexes as-is
+
+    // Get all the enumeration values in the passed-in column
+    std::vector<ValueType> enums_in_write;
+    uint64_t num_elems = value_array->length;
+    if (strcmp(value_schema->format, "b") == 0) {
+        // Specially handle Boolean types as their representation in Arrow (bit)
+        // is different from what is in TileDB (uint8_t)
+        auto casted = _cast_bit_to_uint8(value_schema, value_array);
+        enums_in_write.assign(
+            (ValueType*)casted.data(), (ValueType*)casted.data() + num_elems);
+    } else {
+        // General case
+        const void* data;
+        if (value_array->n_buffers == 3) {
+            data = value_array->buffers[2];
+        } else {
+            data = value_array->buffers[1];
+        }
+        enums_in_write.assign((ValueType*)data, (ValueType*)data + num_elems);
+    }
+
+    // Get all the enumeration values in the on-disk TileDB attribute
+    std::string column_name = index_schema->name;
+    auto enmr = ArrayExperimental::get_enumeration(
+        *ctx_->tiledb_ctx(), *arr_, column_name);
+    std::vector<ValueType> enums_existing = enmr.as_vector<ValueType>();
+
+    // Find any new enumeration values
+    std::vector<ValueType> extend_values;
+    for (auto enum_val : enums_in_write) {
+        if (std::find(enums_existing.begin(), enums_existing.end(), enum_val) ==
+            enums_existing.end()) {
+            extend_values.push_back(enum_val);
+        }
+    }
+
+    // extend_values = {true, false};
+    if (extend_values.size() != 0) {
+        // We have new enumeration values; additional processing needed
+
+        // Check if the number of new enumeration will cause an overflow if
+        // extended
+        auto disk_index_type = tiledb_schema()->attribute(column_name).type();
+        uint64_t max_capacity = _get_max_capacity(disk_index_type);
+        auto free_capacity = max_capacity - enums_existing.size();
+        if (free_capacity < extend_values.size()) {
+            throw TileDBSOMAError(
+                "Cannot extend enumeration; reached maximum capacity");
+        }
+
+        // Take the existing enumeration values on disk and extend with the new
+        // enumeration values
+        auto extended_enmr = enmr.extend(extend_values);
+        se.extend_enumeration(extended_enmr);
+
+        // If the passed-in enumerations are only a subset of the new extended
+        // enumerations, then we will need to remap the indexes. ie. the user
+        // passes in values [B, C] which maps to indexes [0, 1]. However, the
+        // full set of extended enumerations is [A, B, C] which means we need to
+        // remap [B, C] to be indexes [1, 2]
+        _remap_indexes(
+            column_name,
+            extended_enmr,
+            enums_in_write,
+            index_schema,
+            index_array);
+
+        return true;
+    } else {
+        // We do not need to extend the enumeration; write as-is
```

Comment:
Looks like you're losing this bugfix #2897 -- we do need to remap indices even if the enum isn't being extended:

* First write has red,yellow,green at 0,1,2
* Second write has yellow,green at 0,1 needing remap to 1,2

I created a unit-test case for this on #2897 ... 👀

Reply:
Aha, see line 1192 -- this is an old omission on my part, not a new omission on your part -- my #2897 didn't fix both the string and non-string variants of `_extend_and_evolve_schema` but it should have -- and on 2897 I only unit-tested the enum-of-string case ...

I can put up a separate/follow-on PR, to implement my incomplete bugfix, if you like -- or you can do it on this one; I'm fine either way

Reply:
I will take care of it in a follow-up PR.